File: bugsnag-android-core/src/main/java/com/bugsnag/android/Breadcrumb.java
package com.bugsnag.android;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.IOException;
import java.util.Date;
import java.util.Map;

@SuppressWarnings("ConstantConditions")
public class Breadcrumb implements JsonStream.Streamable {

    private final BreadcrumbImpl impl;
    private final Logger logger;

    Breadcrumb(@NonNull String message, @NonNull Logger logger) {
        this.impl = new BreadcrumbImpl(message);
        this.logger = logger;
    }

    Breadcrumb(@NonNull String message,
               @NonNull BreadcrumbType type,
               @Nullable Map<String, Object> metadata,
               @NonNull Date timestamp,
               @NonNull Logger logger) {
        this.impl = new BreadcrumbImpl(message, type, metadata, timestamp);
        this.logger = logger;
    }

    private void error(String property) {
        logger.e("Invalid null value supplied to breadcrumb." + property + ", ignoring");
    }

    public void setMessage(@NonNull String message) {
        if (message != null) {
            impl.setMessage(message);
        } else {
            error("message");
        }
    }

    @NonNull
    public String getMessage() {
        return impl.getMessage();
    }

    public void setType(@NonNull BreadcrumbType type) {
        if (type != null) {
            impl.setType(type);
        } else {
            error("type");
        }
    }

    @NonNull
    public BreadcrumbType getType() {
        return impl.getType();
    }

    public void setMetadata(@NonNull Map<String, Object> metadata) {
        impl.setMetadata(metadata);
    }

    @NonNull
    public Map<String, Object> getMetadata() {
        return impl.getMetadata();
    }

    @NonNull
    public Date getTimestamp() {
        return impl.getTimestamp();
    }

    @Override
    public void toStream(@NonNull JsonStream stream) throws IOException {
        impl.toStream(stream);
    }
}

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Breadcrumb.kt
package com.bugsnag.android

import java.io.IOException
import java.util.Date

/**
 * In order to understand what happened in your application before each crash, it can be helpful
 * to leave short log statements that we call breadcrumbs. Breadcrumbs are
 * attached to a crash to help diagnose what events lead to the error.
 */
internal class BreadcrumbInternal internal constructor(
    var message: String,
    var type: BreadcrumbType,
    var metadata: MutableMap<String, Any?>?,
    val timestamp: Date = Date()
) : JsonStream.Streamable {

    internal constructor(message: String) : this(
        "manual",
        BreadcrumbType.MANUAL,
        mutableMapOf(Pair("message", message)),
        Date()
    )

    @Throws(IOException::class)
    override fun toStream(writer: JsonStream) {
        writer.beginObject()
        writer.name("timestamp").value(DateUtils.toIso8601(timestamp))
        writer.name("name").value(message)
        writer.name("type").value(type.toString())
        writer.name("metaData")
        writer.value(metadata, true)
        writer.endObject()
    }
}
